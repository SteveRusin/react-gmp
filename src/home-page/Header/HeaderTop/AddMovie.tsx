import { ButtonSecondary } from '@shared';

import styled from 'styled-components';

const AddMovieStyled = styled(ButtonSecondary)`
  font-size: 2rem;
`;

export const AddMovie = () => <AddMovieStyled>+ ADD MOVIE</AddMovieStyled>;
