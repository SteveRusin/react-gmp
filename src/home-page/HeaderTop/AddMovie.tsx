import { ButtonSecondary } from '@shared';

export const AddMovie = () => <ButtonSecondary>+ ADD MOVIE</ButtonSecondary>;
