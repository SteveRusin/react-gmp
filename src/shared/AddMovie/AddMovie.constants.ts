import { MovieGenre } from '@api/Movies';

import { SelectOption } from '@shared';

export const GENRE_OPTIONS: SelectOption<MovieGenre>[] = Object.values(
  MovieGenre
).map((genre) => ({
  id: genre,
  name: genre,
}));
