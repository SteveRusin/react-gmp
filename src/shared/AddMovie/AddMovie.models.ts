import { useFormik, FormikProps } from 'formik';

import { Movie } from '@api/Movies';

export interface AddMovieProps {
  movie?: Movie;
  formik: FormikProps<Movie>;
}

export interface AddMovieModalProps {
  movie?: Movie;
  onClose: () => void;
  onSubmit: (movie: Movie) => void;
}

