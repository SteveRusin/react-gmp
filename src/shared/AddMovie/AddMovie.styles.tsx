import styled from 'styled-components';

import { TextArea } from '@shared/FormControls';
import { FlexColumnContainer } from '@shared/FlexColumnContainer';

export const AddMovieModalWrapper = styled(FlexColumnContainer)`
  max-height: 800px;
  width: 950px;
  padding: 30px 30px 60px;

  background-color: ${({ theme }) => theme.palette.accent};
`;

export const ContentWrapper = styled(FlexColumnContainer)``;

export const AddMovieActionsWrapper = styled.div`
  display: flex;
  margin-left: auto;
  gap: 1em;
  font-size: 2rem;
`;

export const AddMovieForm = styled.div`
  display: flex;
  gap: 20px;
  margin-bottom: 20px;
  .form-left {
    width: 80%;
  }

  .form-right,
  .form-left {
    display: flex;
    flex-direction: column;
    gap: 30px;
  }
`;

export const MovieDescription = styled(TextArea)`
  flex: 1;
  margin-bottom: 20px;
  resize: none;
`;
