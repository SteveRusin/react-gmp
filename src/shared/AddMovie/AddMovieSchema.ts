import { object, string, number, array } from 'yup';

export const addMovieSchema = object().shape({
  title: string().required(),
  runtime: number().required(),
  description: string().required(),
  coverUrl: string().url().required(),
  genre: array().required(),
});
