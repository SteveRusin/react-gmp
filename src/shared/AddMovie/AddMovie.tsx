import { FC } from 'react';
import { Formik, useFormik } from 'formik';

import { MovieGenre, Movie } from '@api/Movies';
import { Label, Input } from '@shared/FormControls';
import { Select } from '@shared';

import { AddMovieForm, MovieDescription } from './AddMovie.styles';
import { AddMovieProps } from './AddMovie.models';
import { GENRE_OPTIONS } from './AddMovie.constants';

export const AddMovie: FC<AddMovieProps> = ({ movie, formik }) => {
  const { values, handleChange } = formik;

  return (
    <>
      <AddMovieForm>
        <div className='form-left'>
          <Label>
            TITLE
            <Input
              name='title'
              value={values.title}
              placeholder='Enter movie title'
              onChange={handleChange}
            />
          </Label>
          <Label>
            MOVIE URL
            <Input
              name='coverUrl'
              value={values.coverUrl}
              placeholder='https://'
              onChange={handleChange}
            />
          </Label>
          <Label>
            GENRE
            <Select<MovieGenre>
              multi={true}
              value={values.genre}
              options={GENRE_OPTIONS}
              onSelect={(ids) => formik.setFieldValue('genre', ids)}
              placeholder='Select genre'
            />
          </Label>
        </div>
        <div className='form-right'>
          <Label>
            {/*Should be a datepicker. Use input for now*/}
            RELEASE DATE
            <Input
              readOnly
              name='releaseDate'
              value={values.releaseDate}
              placeholder='Select Date'
            />
          </Label>
          <Label>
            RATING
            <Input
              value={values.rating}
              name='rating'
              placeholder='7.8'
              onChange={handleChange}
            />
          </Label>
          <Label>
            RUNTIME
            <Input
              value={values.runtime}
              name='runtime'
              type="number"
              placeholder='minutes'
              onChange={handleChange}
            />
          </Label>
        </div>
      </AddMovieForm>
      <Label style={{ flex: 1 }}>
        OVERVIEW
        <MovieDescription
          name='description'
          value={values.description}
          onChange={handleChange}
          placeholder='Movie description'
        />
      </Label>
    </>
  );
};
