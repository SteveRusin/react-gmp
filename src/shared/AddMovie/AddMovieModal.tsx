import { PureComponent } from 'react';

import { ButtonPrimary, ButtonPrimaryOutline } from '@shared';
import { CloseModal, ModalTitle, ModalHeader } from '@shared/Modal';

import { AddMovie } from './AddMovie';
import { AddMovieModalProps } from './AddMovie.models';
import {
  AddMovieModalWrapper,
  AddMovieActionsWrapper,
  ContentWrapper,
} from './AddMovie.styles';

export class AddMovieModal extends PureComponent<AddMovieModalProps> {
  render() {
    return (
      <AddMovieModalWrapper>
        <ModalHeader>
          <ModalTitle>
            {this.props.movie ? 'EDIT MOVIE' : 'ADD MOVIE'}
          </ModalTitle>
          <CloseModal size='24' onClick={this.props.onClose} />
        </ModalHeader>
        <ContentWrapper>
          <AddMovie movie={this.props.movie}></AddMovie>
        </ContentWrapper>
        <AddMovieActionsWrapper>
          <ButtonPrimaryOutline>
            RESET
          </ButtonPrimaryOutline>
          <ButtonPrimary onClick={this.props.onSubmit}>SUBMIT</ButtonPrimary>
        </AddMovieActionsWrapper>
      </AddMovieModalWrapper>
    );
  }
}
