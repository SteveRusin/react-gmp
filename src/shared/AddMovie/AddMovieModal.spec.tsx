import { Modal } from '@shared';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@theme';
import { AddMovieModalProps } from './AddMovie.models';
import { AddMovieModal } from './AddMovieModal';
import { render, screen } from '@testing-library/react';

function renderComponent(props: AddMovieModalProps) {
  return (
    <ThemeProvider theme={defaultTheme}>
      <Modal>
        <AddMovieModal {...props} />
      </Modal>
    </ThemeProvider>
  );
}

describe('AddMovie Modal Component', () => {
  let portalRoot: HTMLDivElement;

  beforeEach(() => {
    portalRoot = document.createElement('div');
    portalRoot.setAttribute('id', 'portal-root');
    document.body.appendChild(portalRoot);
  });

  afterEach(() => {
    document.body.removeChild(portalRoot);
  });

  it('should renders correctly', () => {
    render(
      renderComponent({
        onClose: jest.fn(),
        onSubmit: jest.fn(),
      })
    );

    expect(portalRoot).toMatchSnapshot();
  });
});
