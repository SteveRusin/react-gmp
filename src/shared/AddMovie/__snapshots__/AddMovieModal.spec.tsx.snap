// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMovie Modal Component should renders correctly 1`] = `
<div
  id="portal-root"
>
  <div>
    <div
      class="sc-bqyKOL epsUjv"
    >
      <div
        class="sc-bdfBQB sc-fFucqa dzdWDj fIsatb"
      >
        <div
          class="sc-fodVek icvtoC"
        >
          <h2
            class="sc-kstqJO iHJPjr"
          >
            ADD MOVIE
          </h2>
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 iuNRof sc-hBEYId fbvFpy"
            fill="currentColor"
            focusable="false"
            height="24"
            viewBox="0 0 16 16"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.72 3.72a.75.75 0 0 1 1.06 0L8 6.94l3.22-3.22a.75.75 0 1 1 1.06 1.06L9.06 8l3.22 3.22a.75.75 0 1 1-1.06 1.06L8 9.06l-3.22 3.22a.75.75 0 0 1-1.06-1.06L6.94 8 3.72 4.78a.75.75 0 0 1 0-1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="sc-bdfBQB sc-bkzYnD dzdWDj"
        >
          <div
            class="sc-dIUeWJ dDiiVg"
          >
            <div
              class="form-left"
            >
              <label
                class="sc-fubCzh wpnEc"
              >
                TITLE
                <input
                  class="sc-pGacB gfISUd"
                  name="title"
                  placeholder="Enter movie title"
                  value=""
                />
              </label>
              <label
                class="sc-fubCzh wpnEc"
              >
                MOVIE URL
                <input
                  class="sc-pGacB gfISUd"
                  name="coverUrl"
                  placeholder="https://"
                  value=""
                />
              </label>
              <label
                class="sc-fubCzh wpnEc"
              >
                GENRE
                <div
                  class="sc-iqHYmW hotNnh select-component"
                  data-testid="select"
                >
                  <div
                    class="placeholder-wrapper"
                  >
                    <input
                      class="sc-pGacB gfISUd select-trigger"
                      placeholder="Select genre"
                      readonly=""
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="StyledIconBase-ea9ulj-0 iuNRof sc-kEjbQP gZSKvA"
                      color="#F65261"
                      fill="currentColor"
                      focusable="false"
                      height="25"
                      viewBox="0 0 16 16"
                      width="25"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="m4.427 7.427 3.396 3.396a.25.25 0 0 0 .354 0l3.396-3.396A.25.25 0 0 0 11.396 7H4.604a.25.25 0 0 0-.177.427z"
                      />
                    </svg>
                  </div>
                </div>
              </label>
            </div>
            <div
              class="form-right"
            >
              <label
                class="sc-fubCzh wpnEc"
              >
                RELEASE DATE
                <input
                  class="sc-pGacB gfISUd"
                  name="releaseDate"
                  placeholder="Select Date"
                  readonly=""
                  value=""
                />
              </label>
              <label
                class="sc-fubCzh wpnEc"
              >
                RATING
                <input
                  class="sc-pGacB gfISUd"
                  name="rating"
                  placeholder="7.8"
                  value=""
                />
              </label>
              <label
                class="sc-fubCzh wpnEc"
              >
                RUNTIME
                <input
                  class="sc-pGacB gfISUd"
                  name="runtime"
                  placeholder="minutes"
                  type="number"
                  value=""
                />
              </label>
            </div>
          </div>
          <label
            class="sc-fubCzh wpnEc"
            style="flex: 1;"
          >
            OVERVIEW
            <textarea
              class="sc-jrAFXE sc-hHfuMS gratnB dLMlOl"
              name="description"
              placeholder="Movie description"
            />
          </label>
          <div
            class="sc-dmlqKv RVMoI"
          />
        </div>
        <div
          class="sc-idOiZg kWopHX"
        >
          <button
            class="sc-gsTEea sc-jSgvzq jDOibV grpEFO"
          >
            RESET
          </button>
          <button
            class="sc-gsTEea sc-eCstlR jDOibV creYTS"
            type="submit"
          >
            SUBMIT
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
